README
=============================
xgsch2pcb - a GUI for gsch2pcb
==============================


Introduction
------------

When designing a printed circuit board (PCB) it's often desirable to
create a 'schematic' which shows the components to be used and their
connectivity in an abstract fashion.  The connectivity information is
then used to help when designing the actual circuit board.

``gsch2pcb`` is a command-line tool, part of the gEDA suite, which is used
to generate and update a PCB layout.  It works with schematics created
by ``gschem``, part of the gEDA suite, and layouts created by ``pcb``, a
PCB layout system commonly used with gEDA.

``xgsch2pcb`` provides an intuitive, user-friendly graphical interface to
``gsch2pcb``.


Requirements
------------

``xgsch2pcb`` requires:

 * python >= 2.4
 * python-gtk2 >= 2.8
 * python-gobject (if using python-gtk2 >= 2.10)
 * python-dbus
 * A recent version of the gEDA suite
 * A version of ``pcb`` compiled with D-Bus support.

Optionally:
 * python-gnome2 (For launching the about URL)

Building ``xgsch2pcb`` from source code additionally requires:

 * autoconf >= 2.59
 * automake


Installation
------------

If you have fetched the sources from the git repository you will need to
create the necessary autoconf files. To do this, change into the source
directory and execute the command:

  ./autogen.sh

This generates a ./configure script amongst other things, which you can run
to configure the software.

As long as you have all the dependencies installed:

  ./configure && make install


Usage
-----

  xgsch2pcb [project]

The main window is divided into three main areas:

* The toolbar at the top offers the usual options to quit the
  program and to load and save project files.

* The left hand 'Schematic' frame shows a list of schematic pages that
  the PCB layout will be based on. The 'Edit schematic' and 'Edit
  attributes' buttons respectively launch ``gschem`` and ``gattrib`` to
  edit the selected schematic page.

* The right hand 'Layout' pane shows the name of the PCB layout file
  associated with the project. The 'Edit layout' button launches
  ``pcb`` to edit a file, and will offer to update your PCB layout if
  necessary.  The 'Update layout' button forces an update of the PCB
  layout even if one isn't strictly necessary.

In order for the update process to work properly, you should never
launch ``pcb`` to edit your PCB layout except through ``xgsch2pcb``.
If you are editing the layout without ``xgsch2pcb`` running and then
wish to edit your layout you should close ``pcb``, launch
``xgsch2pcb`` and then use the 'Edit layout' button to launch ``pcb``
again.

The update process will carry out the following actions to modify your
layout, after launching ``pcb`` if isn't already running:

1. Remove any elements from the layout that are not in the schematic.

2. Find any elements that are in the schematic but not in the layout,
   and add them to the layout (in the top left corner). N.b. that it's
   probably a good idea to leave this corner of your layout clear
   until the layout is more or less finalised, to avoid new elements
   interfering with elements which have already been placed and
   routed.

3. Clear your rats and load a new rats nest.

4. Update the component pin names to match the pin names on the
   schematic symbol.

Note that the update process won't modify your PCB file on disk, and 
will take into account any changes you have made since you last 
saved.

.. 
   Local Variables:
   mode:rst
   End:


ChangeLog
# Do not edit this file - generated from version control history


commit 26500fa30839f1b6f5fc0b21be29973d84625022
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Mon Sep 28 19:30:16 2009 +0100

    Bump version to 0.1.3

commit 0dc1bbb6076fc380ce718e057aba622ce45827d8
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Mon Sep 28 19:26:30 2009 +0100

    Add translations from Launchpad
    
    Launchpad translations are made under the BSD license, see:
    https://help.launchpad.net/Translations/LicensingFAQ

commit e4167391f1cb1b127278a9dc22849051d40b99c0
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Tue Mar 10 16:18:47 2009 +0000

    Handle blank lines in the project file correctly.
    
    Blank project file lines split up to give a zero length array
    of tokens. Don't try to access the array if it is empty.

commit 386a2fd3bc971a1b42695b930c3e7c3930d13dbd
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Tue Mar 10 16:00:39 2009 +0000

    Pick the correct PCB instance when multiple are open
    
    Instead of breaking out of the loop over PCB instances when we found
    the right one  we continue'd looping. Whoops!

commit 4d571037066113c9d372a7bf71560dcd9cd988e5
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Sun Jul 27 14:53:56 2008 +0100

    Make double-clicking / activating a schematic open it in gschem

commit 8fe2dae043955e3dd36a02c091527befa0624430
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Sun Jun 1 13:45:38 2008 +0100

    Ignore unknown config file options, and re-emit them when saving
    
    This allows .gsch2pcb file options which aren't directly supported
    by xgsch2pcb (including comments) to be read and re-saved without
    breaking the file.

commit 18e9c086ba540ba10e9ed358c2a4b738ba2526f6
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Fri Feb 1 13:01:10 2008 +0000

    Rename translations to drop un-needed ISO-3166 country codes.
    
    From Debian bug #336796:
    
    Using a country part in a translation file is a discouraged practice
    except in very few cases (such as pt_BR).
    
    Using a fr_FR.po file instead of a fr.po file prevents users of fr_CA,
    fr_BE, fr_LU, fr_CH and all other existing and future locales for
    French to benefit from the French translation of the program.
    
    In general PO files should only be named after the
    ISO_639 code of the given language and should not use a country part
    with a ISO-3166 code. The only accepted expcetions to this are:
    
    -pt_BR for Brazilian Portuguese and pt alone for "standard Portuguese"
    -zh_CN for "Simplified Chinese" use in mailand China and Singapore
    -zh_TW for "Traditional Chinese" used in Taiwan

commit 25bb8e76f8851a4e60f21a44e88462c5c5f57722
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Fri Jan 11 05:20:34 2008 +0000

    Update ChangeLog for the 0.1.2 release

commit 892f310341456407bcc948769cea27754b4fbd40
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Fri Jan 11 05:16:19 2008 +0000

    Bump version to 0.1.2

commit 582528e8901228714f608f71792469fea00f0a1a
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Fri Jan 11 05:15:53 2008 +0000

    Ran make update-po in all po directories in prep for a new release.

commit d118b25782b51e0e3c490c466ad87f3de0ba9951
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Thu Jan 10 02:39:39 2008 +0000

    Added --with-xdgdatadir configure option
    
    This options sets where icons and menu entries are installed. It defaults
    to ${datadir}, but may be over-ridden to $HOME/.local/share/ for a "this
    user only" install.

commit 0e745ded110615a4dbaf96039fd4dc9c0ab503fd
Author: Carlos Nieves Onega <cnieves@iespana.es>
Date:   Mon Jan 7 20:47:12 2008 +0100

    Add es_ES.po to the LINGUAS file, for the new spanish translation.

commit e4931eba3693a6ed61a8293059321015f956aa08
Author: Carlos Nieves Onega <cnieves@iespana.es>
Date:   Mon Jan 7 20:33:16 2008 +0100

    Changed installation instructions for people using git repository.
    
    Auto-generated files where removed from the repository, so the user has to
    create them using ./autogen.sh script.

commit e5803af6b76acbd9662611cb87cddd50cd9f3f10
Author: Carlos Nieves Onega <cnieves@iespana.es>
Date:   Mon Jan 7 20:31:30 2008 +0100

    Added new spanish (Spain) translation.

commit 83ecf6649201778722c7679bfe561e7794976cca
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Sat Jan 5 18:11:46 2008 +0000

    Update "translator-credit" in the German translation

commit 59ad0b4d0e3d3be3a790487c0d5e087894299da2
Author: Werner Hoch <werner.ho@gmx.de>
Date:   Thu Jan 3 21:23:40 2008 +0100

    initial german translation de_DE.po
    
    translated all messages
    added de_DE to the LINGUAS file

commit 40fb4865cef90a51ee388a088f4fa495e56b17a8
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Wed Jan 2 19:17:37 2008 +0000

    Prefix "geda-" to icon names to help namespace the icon theme.

commit 8a5456bd8a6b9934b9a2c583e03bc4b7baf21a11
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Sun Dec 30 03:38:51 2007 +0000

    Remove auto-generated geda-xgsch2pcb.pot file from repository

commit ccdb60fa83b81051df1d9a84c22c379a55770ef3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Sun Dec 30 03:36:39 2007 +0000

    Remove po/POTFILES.skip now the file it lists is no longer present

commit 278784f4980deb9bb69682ac0d1ff914fb293a47
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Wed Dec 26 22:18:20 2007 +0000

    Update "translator-credit" in the Dutch translation

commit 654fd7dc8279f723fab891c1bae19e3cf1e5d6b7
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Wed Dec 26 21:52:46 2007 +0000

    Add Dutch translation

commit eac8acc49f010f027aa46e45fd6957964dab31a3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Tue Dec 25 05:07:35 2007 +0000

    Remove autogenerated files from source control

commit 7f6d5f1fc2a03db3238a8f5d1f800e3247cfadc3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   Tue Dec 25 04:58:28 2007 +0000


update-changelogs.sh
#!/bin/bash
#
# ./update-changelogs.sh <options>
# Updates ChangeLog files.
#
# e.g. if the main gEDA branch is called master, then to update all
# ChangeLogs with revisions since the switch to git, do:
#
# ./update-changelogs.sh 1.0-20070526..master
#
# This should be only be run while preparing a release, in order to
# ensure that the released tarballs contain change logs which can be
# viewed without access to the git repository.
#
#

GCLHEADER="# Do not edit this file - generated from version control history"
GCLFOOTER="#GCL#"

if ! git log -n1 > /dev/null; then
    exit $?
fi

for cl in $(find . -name ChangeLog); do
    echo "Updating ${cl}"
    dir=$(dirname $cl)
    if tail -n1 ${cl} | grep $GCLFOOTER > /dev/null; then
        echo -e "$GCLHEADER\n\n" > $cl
        git log --pretty=medium $@ -- $dir >> $cl
        echo -e "\n\n$GCLFOOTER" >> $cl
    else
        echo "${cl}: Doesn't appear to be git format log, skipping"
    fi
done


lib/xgsch2pcb/funcs.py
# -*-Python-*-

# xgsch2pcb - a GUI for gsch2pcb
# Copyright (C) 2006 University of Cambridge
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import commands
import os

def find_tool_path (tool):

    which = commands.getstatusoutput("which %s" % tool)
    if which[0]:
        return None
    
    return which[1]
    
def rel_path(topath, fromdir=""):

    sep = os.path.sep
    pardir = os.path.pardir

    # Expand paths:
    # "" becomes the current dir
    # /foo/../ sections are contracted
    fromdir = os.path.abspath( fromdir )
    topath = os.path.abspath( topath )

    if fromdir == topath:
        return ''

    # We need a directory separator at the end for consistency
    if not fromdir.endswith( sep ):
        fromdir = fromdir + sep

    # Compute the common prefix of the paths
    # Note, that this is done on a string basis:
    # /home and /homedir share a common prefix of /home
    prefix = os.path.commonprefix([fromdir, topath])
    
    # If the topath doesn't have fromdir as a common part: 
    if (prefix != fromdir):

        # Go down one directory in the fromdir, and prepend
        # a relative move down one dir in the output string
        fromdir = os.path.abspath( fromdir + pardir )

        # Recurse down
        return pardir + sep + rel_path(fromdir, topath)
    
    # Return the topath without the common prefix
    return topath[len(prefix):]



lib/xgsch2pcb/templates.py
# -*-Python-*-

# xgsch2pcb - a GUI for gsch2pcb
# Copyright (C) 2006 University of Cambridge
# Copyright (C) 2007 Peter Clifton <pcjc2@cam.ac.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import os
import shutil

# xgsch2pcb-specific modules
import config
from gsch2pcbproject import Gsch2PCBProject

def template_path( template, file ):
    path = os.path.join( config.templatesdir, template )
    path = os.path.join( path, file )
    return path


def replace_project_name( file, project ):
    """
Produce the filename which would exist for a given project name.

This implementation simply replaces the first occurance of the string
'template' with the project name.

It should only be called with relative paths, otherwise any prefix directory
containing the string template (e.g. the templates directory!) would be
replaced instead of the desired portion of the file name.
    """
    return file.replace( 'template', project, 1 )


def list_templates():

    template_list = []

    try:
        filelist = os.listdir( config.templatesdir )
    except:
        print "Couldn't list templates directory"
        return template_list

    filelist.sort()

    for template in filelist:
        try:
            templ = gsch2pcb_template( template )
            [name, description] = templ.read_description()
            template_list.append( [template, name, description] )
        except:
            print "Couldn't read a template in dir " + template

    return template_list


class gsch2pcb_template:

    def __init__(self, template):
        self.template = template
        self.filename = template_path( template, 'template.gsch2pcb')

        # Load the project into memory
        self.template_project = Gsch2PCBProject( self.filename )

    def read_description(self):

        file = open( template_path( self.template, 'template.txt' ), 'r' )
        name = file.readline().strip()
        blank = file.readline()
        if blank != "\n":
            print "Invalid file format for this template"
            file.close()
        # Join remaining lines in the file to a single string
        description = "".join( file.readlines() )
        file.close()
        return [name, description]


    def would_create(self, projectname):
        """Returns a list of files which this templates would create"""
        # TODO: Should this be a relative or absolute path??

        filelist = []

        # Would create a new project file:
        # TODO: Should the template have this as a relative path to some dir?
        filelist.append( os.path.basename( self.template_project.filename ) )
        # Would create the output file:
        # TODO: REMOVE HARDCODED EXTENSION
        filelist.append( self.template_project.output_name + '.pcb' )
        # would create the pages:
        filelist.extend( self.template_project.pages )

        # Substitute the templatized filenames for ones matching the new project name
        new_filelist = []
        for file in filelist:
          new_filelist.append( replace_project_name( file, projectname ) )

        return new_filelist


    def apply(self, projectname):

        # Copy the output file
        output_name = self.template_project.output_name
        new_output_name = replace_project_name( output_name, projectname )
        output_file = output_name + '.pcb'         # TODO: REMOVE HARDCODED EXTENSION
        new_output_file = new_output_name + '.pcb' # TODO: REMOVE HARDCODED EXTENSION
        shutil.copy( template_path( self.template, output_file ), new_output_file )

        # Create a new gsch2pcb project in the new location, with the new output filename
        new_project = Gsch2PCBProject( None, new_output_name )

        # Copy schematic pages, and add to the new project
        for page_file in self.template_project.pages:
            new_page_file = replace_project_name( page_file, projectname )
            shutil.copy( template_path( self.template, page_file ), new_page_file )
            new_project.add_page( new_page_file )

        # TODO: We could just hand off the new project file without saving of course?
        # TODO: REMOVE HARDCODED EXTENSION
        new_project.save(projectname + '.gsch2pcb')



lib/xgsch2pcb/new_project_gui.py
# -*-Python-*-

# xgsch2pcb - a GUI for gsch2pcb
# Copyright (C) 2006 University of Cambridge
# Copyright (C) 2007 Peter Clifton <pcjc2@cam.ac.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import gtk, gtk.gdk, gobject

import config

import gettext
t = gettext.translation(config.PACKAGE, config.localedir, fallback=True)
_ = t.ugettext

# xgsch2pcb-specific modules
from templates import *
from assistant import *

class NewProjectAssistant(Assistant):

    __gsignals__ = { 'project-apply' :
                            ( gobject.SIGNAL_NO_RECURSE,
                              gobject.TYPE_NONE,
                              (gobject.TYPE_STRING, )),
                   }

    template = None

    def assistant_apply(self, assistant):

        # TODO: Why special case testing the writability of just this ONE output file?
        filename = self.get_filename()

        # Create a new zero-length file
        try:
            if self.template:
                # Apply any chosen template
                templ = gsch2pcb_template( self.template )
                templ.apply( self.get_projectname() )
            else:
                # Otherwise just write out a new project file
                # Start with a filename of None to ensure that
                # we don't load an existing project file.
                new_project = Gsch2PCBProject( None,
                                               self.get_projectname() )

                new_project.save(self.get_filename())
        except IOError, (errno, strerror):
            md = gtk.MessageDialog(self,
                                   (gtk.DIALOG_MODAL |
                                    gtk.DIALOG_DESTROY_WITH_PARENT),
                                   gtk.MESSAGE_ERROR,
                                   gtk.BUTTONS_OK )

            md.set_markup( _('<span weight="bold" size="larger">Could not create project</span>\n\nError %i: %s') % (errno, strerror) )
            md.show_all()
            md.run()
            md.hide_all()
            return
        except:
            md = gtk.MessageDialog(self,
                                   (gtk.DIALOG_MODAL |
                                    gtk.DIALOG_DESTROY_WITH_PARENT),
                                   gtk.MESSAGE_ERROR,
                                   gtk.BUTTONS_OK )

            md.set_markup( _('<span weight="bold" size="larger">Could not create project</span>') )
            md.show_all()
            md.run()
            md.hide_all()
            return

        self.emit('project-apply', filename)


    def assistant_cancel(self, assistant):
        self.destroy()

    def assistant_close(self, assistant):
        self.destroy()

    def template_radio_toggled(self, button):
        is_blank = self.blankradio.get_active()
        self.templateview.set_sensitive(not is_blank)
        if is_blank:
            self.set_page_complete(self.template_page, True)
        else:
            # Synthesise a selection changed update on the tree-view
            self.template_selection_changed(self.templateview.get_selection())

    def template_selection_changed(self, treeselection):
        [model,iters] = treeselection.get_selected()
        if iters:
            self.set_page_complete(self.template_page, True)
            template_no = model.get_path(iters)[0]
            [self.template, description] = model.get(iters,0,2)   # TODO: Remove magic numbers
            self.description.set_text(description)
        else:
            self.set_page_complete(self.template_page, False)
            self.template = None
            self.description.set_text("")

    def get_path(self):
        path = self.filebutton.get_filename()
        return path

    def get_filename(self):
        filename = self.filename.get_text()
        # TODO: REMOVE HARDCODED EXTENTIONS - SHOULD THIS CODE GO HERE?
        if not filename.endswith('.gsch2pcb'):
            filename += '.gsch2pcb'
        return filename

    def get_projectname(self):
        # TODO: Decide if this is the right way to do this
        filename = self.get_filename()
        # TODO: REMOVE HARDCODED EXTENTIONS - SHOULD THIS CODE GO HERE?
        projectname = filename[0:filename.rfind( '.gsch2pcb' )]
        return projectname

    def __init__(self, parent):
        Assistant.__init__(self)
        self.set_transient_for( parent )
        self.set_position( gtk.WIN_POS_CENTER_ON_PARENT )

        # Render a stock "NEW" icon to display on the pages
        image = self.render_icon( gtk.STOCK_NEW, gtk.ICON_SIZE_DIALOG )

        # ====================
        # Choose template page
        # ====================

        page = gtk.VBox()
        page.set_border_width(12)
        page.set_spacing(6)
        label = gtk.Label(_("<b>Choose project template</b>"))
        label.set_use_markup(True)
        label.set_alignment(0, 0.5)
        page.pack_start(label, False, False);

        align = gtk.Alignment(0, 0, 1, 1)
        page.pack_start(align, True, True)
        align.set_padding(12,12,12,12)
        options = gtk.VBox()
        align.add(options)
        options.set_spacing(6)
        self.blankradio = gtk.RadioButton(label=_("Blank"))
        self.templradio = gtk.RadioButton(group=self.blankradio,
                                          label=_("From template:"))
        options.pack_start(self.blankradio, False, False)
        options.pack_start(self.templradio, False, False)

        self.templatelist = gtk.ListStore(gobject.TYPE_STRING, gobject.TYPE_STRING, gobject.TYPE_STRING)
        templates = list_templates()
        for template in templates:
            self.templatelist.append( template )

        if len(templates) == 0:
            self.templradio.set_sensitive(False)
            self.templatelist.append( ['', _("(No templates found)"),''] )

        textrenderer = gtk.CellRendererText()
        textcol = gtk.TreeViewColumn(None, textrenderer, text=1)

        self.templateview = gtk.TreeView(self.templatelist)
        self.templateview.append_column(textcol)
        self.templateview.set_headers_visible(False)
        self.templateview.set_sensitive(False)

        scrollwin = gtk.ScrolledWindow()
        scrollwin.set_policy(gtk.POLICY_NEVER, gtk.POLICY_ALWAYS)
        scrollwin.set_shadow_type(gtk.SHADOW_IN)
        scrollwin.add(self.templateview)

        align = gtk.Alignment(0, 0, 1, 1)
        align.set_padding(0, 0, 18, 0)
        align.add(scrollwin)
        options.pack_start(align, True, True)

        self.description = gtk.Label()

        self.description.set_line_wrap(True)
        self.description.set_alignment(0, 0)
        self.description.set_padding(18, 0)
        self.description.set_selectable(True)
        self.description.set_max_width_chars(0)
        options.pack_start(self.description, False, False)

        self.append_page(page)
        self.set_page_title(page, _("Create new project"))

        self.set_page_side_image(page,image)

        self.set_page_type(page, ASSISTANT_PAGE_CONTENT)
        self.set_page_complete(page, True)
        self.template_page = page

        self.blankradio.connect('toggled', self.template_radio_toggled)
        self.templradio.connect('toggled', self.template_radio_toggled)

        treeselection = self.templateview.get_selection()
        treeselection.connect('changed', self.template_selection_changed)

        # ============================
        # Choose project filename page
        # ============================

        page = gtk.VBox()
        page.set_border_width(12)
        page.set_spacing(6)
        label = gtk.Label(_("<b>Choose project filename</b>"))
        label.set_use_markup(True)
        label.set_alignment(0, 0.5)
        page.pack_start(label, False, False);

        align = gtk.Alignment(0, 0, 1, 1)
        page.pack_start(align, True, True)
        align.set_padding(12,12,12,12)

        options = gtk.VBox()
        align.add(options)

        table = gtk.Table(2,2)
        table.set_col_spacings( 6 ) # TODO: Remove magic numbers
        table.set_row_spacings( 6 ) # TODO: Remove magic numbers
        label = gtk.Label(_("Project name:"))
        label.set_alignment(0, 0.5)
        table.attach(label, 0, 1, 0, 1, gtk.FILL, 0)
        self.filename = gtk.Entry()
        table.attach(self.filename, 1, 2, 0, 1, gtk.EXPAND | gtk.FILL, 0)
        label = gtk.Label(_("Location:"))
        label.set_alignment(0, 0.5)
        table.attach(label, 0, 1, 1, 2, gtk.FILL, 0)

        def filebutton_selection_changed_cb(filechooser):
            # Change to the specified location
            os.chdir( self.get_path() )

        self.filebutton = gtk.FileChooserButton(_('Select project location...'))
        self.filebutton.connect( "selection-changed", filebutton_selection_changed_cb )
        self.filebutton.set_local_only(True)
        self.filebutton.set_action(gtk.FILE_CHOOSER_ACTION_SELECT_FOLDER)

        table.attach(self.filebutton, 1, 2, 1, 2, gtk.FILL | gtk.EXPAND, 0)

        options.pack_start(table, False, False)

        self.append_page(page)
        self.set_page_title(page, _("Create new project"))
        self.set_page_side_image(page,image)
        self.set_page_type(page, ASSISTANT_PAGE_CONTENT)
        self.set_page_complete(page, False)
        self.filename_page = page

        def filename_changed_cb( filename_entry ):
            entrytext = self.filename.get_text()
            self.set_page_complete(self.filename_page, (not entrytext == ""))

        self.filename.connect('changed', filename_changed_cb)

        # ================
        # Creation summary
        # ================

        page = gtk.VBox()
        page.set_border_width(12)
        page.set_spacing(6)
        label = gtk.Label(_("<b>Project summary</b>"))
        label.set_use_markup(True)
        label.set_alignment(0, 0.5)
        page.pack_start(label, False, False);

        align = gtk.Alignment(0, 0, 1, 1)
        page.pack_start(align, True, True)
        align.set_padding(12,12,12,12)

        explanation = gtk.VBox()
        align.add(explanation)

        self.newfiles_frame = gtk.Frame(_("<b>New files to be created:</b>"))
        self.newfiles_frame.get_label_widget().set_use_markup(True)
        self.newfiles_frame.set_shadow_type(gtk.SHADOW_NONE)
        explanation.pack_start(self.newfiles_frame, False, False)

        align = gtk.Alignment(0, 0, 1, 1)
        align.set_padding(0,12,12,12)
        self.newfiles_frame.add(align)

        self.newfiles_list = gtk.Label()
        self.newfiles_list.set_alignment(0, 0.5)
        self.newfiles_list.set_padding(0,12)
        align.add(self.newfiles_list)

        self.overwrite_frame = gtk.Frame(_("<b>The following files would be overwritten:</b>"))
        self.overwrite_frame.get_label_widget().set_use_markup(True)
        self.overwrite_frame.set_shadow_type(gtk.SHADOW_NONE)
        explanation.pack_start(self.overwrite_frame, False, False)

        align = gtk.Alignment(0, 0, 1, 1)
        align.set_padding(0,12,12,12)
        self.overwrite_frame.add(align)

        vbox = gtk.VBox()
        align.add(vbox)

        self.overwrite_list = gtk.Label()
        self.overwrite_list.set_alignment(0, 0.5)
        self.overwrite_list.set_padding(0,12)
        vbox.pack_start(self.overwrite_list, False, False)

        self.confirm_overwrite = gtk.CheckButton(_("Confirm overwrite"))
        vbox.pack_start(self.confirm_overwrite, False, False)

        def confirm_overwrite_toggled_cb( togglebutton ):
            confirmed = togglebutton.get_active()
            self.set_page_complete(self.summary_page, confirmed)

        self.confirm_overwrite.connect( 'toggled', confirm_overwrite_toggled_cb )

        self.append_page(page)
        self.set_page_title(page, _("Create new project"))
        self.set_page_side_image(page,image)
        self.set_page_type(page, ASSISTANT_PAGE_CONFIRM)
        self.summary_page = page


        def check_overwrites():
            if self.template:
                templ = gsch2pcb_template( self.template )
                file_list = templ.would_create( self.get_projectname() )
            else:
                file_list = [self.get_filename()]
            newfiles_list = []
            overwrite_list = []
            for file in file_list:
                if os.path.exists( file ):
                    overwrite_list.append( file )
                else:
                    newfiles_list.append( file )
            return [newfiles_list, overwrite_list]

        def assistant_prepare_cb(assistant, page):
            if page is self.summary_page:
                # Summary page before creating the new project on disk
                [newfiles_list, overwrite_list] = check_overwrites()
                self.newfiles_list.set_text( '\n'.join( newfiles_list ) )
                self.overwrite_list.set_text( '\n'.join( overwrite_list ) )

                no_newfiles = (newfiles_list == [])
                if no_newfiles:
                    self.newfiles_frame.hide_all()
                else:
                    self.newfiles_frame.show_all()

                no_overwrite = (overwrite_list == [])
                if no_overwrite:
                    # No files will be overwritten, we are done
                    self.overwrite_frame.hide_all()
                    self.set_page_complete(self.summary_page, True)
                else:
                    # Need confirmation before overwriting files
                    self.overwrite_frame.show_all()
                    self.confirm_overwrite.set_active(False)

        # GtkAssistant signals
        self.connect('prepare', assistant_prepare_cb  )
        self.connect('apply',   self.assistant_apply  )
        self.connect('cancel',  self.assistant_cancel )
        self.connect('close',   self.assistant_close  )


gobject.type_register( NewProjectAssistant )
